File: LeetCode/Python Solutions/Maximum Running Time of N Computers/maximum.py
class Solution:
    def maxRunTime(self, n: int, batteries: List[int]) -> int:
        # Calculate the total power of all batteries
        total_power = sum(batteries)
        
        # The lower and upper bounds of our search space
        left, right = 1, total_power // n

        # Binary search for the maximum possible running time
        while left < right:
            # Calculate the middle of the current search space
            time = (left + right + 1) // 2
            
            # Check if the current running time is possible with the given batteries
            if self.check(batteries, n, time):
                # If it is possible, we update the lower bound to the current time
                left = time
            else:
                # If it is not possible, we update the upper bound to the current time minus 1
                right = time - 1
                
        # Return the maximum possible running time
        return left
    
    def check(self, batteries: List[int], n: int, time: int) -> bool:
        # Calculate the total power that can be provided for the given running time
        total_power_for_time = sum(min(battery, time) for battery in batteries)
        
        # Check if the total power is at least the product of the running time and the number of computers
        return total_power_for_time >= time * n

File: LeetCode/Java Solutions/Maximum Running Time of N Computers/maximum.java
public class Solution {
    // Function to calculate the maximum runtime for all computers
    public long maxRunTime(int numberOfComputers, int[] batteries) {
        // Calculate the total power of all batteries
        long totalBatteryPower = 0;
        for (int batteryPower : batteries) {
            totalBatteryPower += batteryPower;
        }
        
        // Set the initial search bounds for binary search
        long lowerBound = 1, upperBound = totalBatteryPower / numberOfComputers;
        
        // Perform binary search to find the maximum possible time
        while (lowerBound < upperBound) {
            // Calculate the midpoint of the current search range
            long midpointTime = (lowerBound + upperBound + 1) / 2;
            
            // Check if it's possible to run the computers for the midpoint time
            if (isTimePossible(batteries, numberOfComputers, midpointTime)) {
                // If it's possible, update the lower bound to the midpoint time
                lowerBound = midpointTime;
            } else {
                // If it's not possible, update the upper bound to one less than the midpoint time
                upperBound = midpointTime - 1;
            }
        }
        
        // Return the maximum possible time
        return lowerBound;
    }
    
    // Function to check if it's possible to run all computers for a given time
    public boolean isTimePossible(int[] batteries, int numberOfComputers, long time) {
        // Calculate the total available power for the given time
        long totalAvailablePower = 0;
        for(int batteryPower : batteries) {
            // Each battery can contribute its power or the time, whichever is smaller
            totalAvailablePower += Math.min(batteryPower, time);
        }
        
        // Check if the total available power is enough to run all computers for the given time
        return (totalAvailablePower >= (long) time * numberOfComputers);
    }
}

